app/src/main/java/org/simple/clinic/overdue/AppointmentCancelReason.kt

```diff
@@ -0,0 +1,74 @@
+package org.simple.clinic.overdue
+
+import android.arch.persistence.room.TypeConverter
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.ToJson
+
+sealed class AppointmentCancelReason {
+
+  object PatientNotResponding : AppointmentCancelReason()
+
+  object Moved : AppointmentCancelReason()
+
+  object Dead : AppointmentCancelReason()
+
+  object Other : AppointmentCancelReason()
+
+  data class Unknown(val actualValue: String) : AppointmentCancelReason()
+
+  /**
+   * It'll be nice to write a code generator instead of hand-typing
+   * and manually maintaining this adapter in the future.
+   */
+  object TypeAdapter {
+    val KNOWN_MAPPINGS = mapOf(
+        PatientNotResponding to "not_responding",
+        Moved to "moved",
+        Dead to "dead",
+        Other to "other")
+
+    fun toEnum(reason: String?): AppointmentCancelReason? {
+      if (reason == null) {
+        return null
+      }
+
+      val foundEnum = KNOWN_MAPPINGS.entries
+          .find { (_, jsonKey) -> jsonKey == reason }
+          ?.key
+      return foundEnum ?: Unknown(actualValue = reason)
+    }
+
+    fun fromEnum(reason: AppointmentCancelReason?): String? {
+      if (reason == null) {
+        return null
+      }
+
+      return if (reason is Unknown) {
+        reason.actualValue
+      } else {
+        KNOWN_MAPPINGS[reason] ?: throw AssertionError("Unknown reason enum: $reason")
+      }
+    }
+
+  }
+
+  class MoshiTypeConverter {
+    @FromJson
+    fun toEnum(reason: String?): AppointmentCancelReason? = TypeAdapter.toEnum(reason)
+
+    @ToJson
+    fun fromEnum(reason: AppointmentCancelReason?): String? = TypeAdapter.fromEnum(reason)
+  }
+
+  class RoomTypeConverter {
+    @TypeConverter
+    fun toEnum(reason: String?): AppointmentCancelReason? = TypeAdapter.toEnum(reason)
+
+    @TypeConverter
+    fun fromEnum(reason: AppointmentCancelReason?): String? = TypeAdapter.fromEnum(reason)
+  }
+
+  companion object {
+    fun random() = TypeAdapter.KNOWN_MAPPINGS.entries.shuffled().first().key
```

Comment:
What do you think of keeping this function here? It's only used in the test and shouldn't belong here, but the mappings are private to this class.

Reply:
I think it's fine for now. Perhaps, in the future, we can investigate using `kotlin-reflect` in tests to do something similar like we have done for enums.

Reply:
Do you want to add a `@VisibleForTesting` annotation to the method?

Reply:
Yes, good idea